mllib/src/main/scala/org/apache/spark/mllib/clustering/LDAModel.scala

```diff
@@ -267,32 +304,42 @@ class LocalLDAModel private[clustering] (
     // by topic (columns of lambda)
     val Elogbeta = LDAUtils.dirichletExpectation(lambda.t).t
 
-    var score = documents.filter(_._2.numNonzeros > 0).map { case (id: Long, termCounts: Vector) =>
-      var docScore = 0.0D
+    // Sum bounds for each document:
+    //  bounds for prob(tokens) + bounds for prob(document-topic distribution)
+    documents.filter(_._2.numNonzeros > 0).map { case (id: Long, termCounts: Vector) =>
+      var docBound = 0.0D
       val (gammad: BDV[Double], _) = OnlineLDAOptimizer.variationalTopicInference(
         termCounts, exp(Elogbeta), brzAlpha, gammaShape, k)
       val Elogthetad: BDV[Double] = LDAUtils.dirichletExpectation(gammad)
 
       // E[log p(doc | theta, beta)]
       termCounts.foreachActive { case (idx, count) =>
-        docScore += count * LDAUtils.logSumExp(Elogthetad + Elogbeta(idx, ::).t)
+        docBound += count * LDAUtils.logSumExp(Elogthetad + Elogbeta(idx, ::).t)
       }
-      // E[log p(theta | alpha) - log q(theta | gamma)]; assumes alpha is a vector
-      docScore += sum((brzAlpha - gammad) :* Elogthetad)
-      docScore += sum(lgamma(gammad) - lgamma(brzAlpha))
-      docScore += lgamma(sum(brzAlpha)) - lgamma(sum(gammad))
+      // E[log p(theta | alpha) - log q(theta | gamma)]
+      docBound += sum((brzAlpha - gammad) :* Elogthetad)
+      docBound += sum(lgamma(gammad) - lgamma(brzAlpha))
+      docBound += lgamma(sum(brzAlpha)) - lgamma(sum(gammad))
 
-      docScore
+      docBound
     }.sum()
+  }
 
-    // E[log p(beta | eta) - log q (beta | lambda)]; assumes eta is a scalar
-    score += sum((eta - lambda) :* Elogbeta)
-    score += sum(lgamma(lambda) - lgamma(eta))
-
+  /**
+   * Bound for prob(topic-term distributions):
+   *   E[log p(beta | eta) - log q(beta | lambda)]
+   *
+   * See Equation (16) in original Online LDA paper, as well as Appendix A.3 in the JMLR version of
+   * the original LDA paper.
+   * @param eta topic-word Dirichlet prior parameter
+   * @param lambda parameters for variational q(beta | lambda) topic-word distributions
+   */
+  private def topicsBound(eta: Double, lambda: BDM[Double]): Double = {
```

Comment:
Did we separate this out just to be consistent with DistributedLDAModel? I found it a bit confusing since the original calculation follows the factorization given in the paper (Eq (3) https://www.cs.princeton.edu/~blei/papers/HoffmanBleiBach2010b.pdf)